index.ts

```diff
@@ -1,165 +1,237 @@
-    .join('\n');
+// EVENTLOG UTILITY FUNCTIONS
+// From the event log, get the transcription so far
+const getLastTranscriptionEvent = (): TranscriptionEvent => {
+  const transcriptionEvents = eventlog.events.filter(e => e.eventType === 'transcription');
+  return transcriptionEvents[transcriptionEvents.length - 1] as TranscriptionEvent;
 }
+const getCutTimestamp = (): number => {
+  const cutTranscriptionEvents = eventlog.events.filter(e => e.eventType === 'cutTranscription');
+  const lastCut = cutTranscriptionEvents.length > 0 ? cutTranscriptionEvents[cutTranscriptionEvents.length - 1].data.lastAudioByteEventTimestamp : eventlog.events[0].timestamp;
+  const responseReflexEvents = eventlog.events.filter(e => e.eventType === 'responseReflex');
+  const lastResponseReflex = responseReflexEvents.length > 0 ? responseReflexEvents[responseReflexEvents.length - 1].timestamp : eventlog.events[0].timestamp;
+  return Math.max(lastResponseReflex, lastCut);
 
-//  INITIALIZE STATE
-const conversation: Conversation = {
-  audioBuffer: Buffer.alloc(0),
-  transcriptions: [],
-  canonicalDialogue: [],
-  watermark: 0,
+}
+const getTransciptionSoFar = (): string => {
+  const cutTranscriptionEvents = eventlog.events.filter(e => e.eventType === 'cutTranscription');
+  const lastTranscriptionEvent = getLastTranscriptionEvent();
+  const lastCutTranscriptionEvent = cutTranscriptionEvents[cutTranscriptionEvents.length - 1];
+  let transcription = cutTranscriptionEvents.map(e => e.data.transcription).join(' ');
+  if (!lastCutTranscriptionEvent || lastCutTranscriptionEvent.timestamp !== lastTranscriptionEvent.timestamp) {
+    transcription = transcription + (lastTranscriptionEvent?.data?.transcription || '')
+  }
+  return transcription
 }
 
-conversation.canonicalDialogue.push({ speaker: 'agent', response: "Hey yacine! long time no see. How have you been?" });
-conversation.canonicalDialogue.push({ speaker: 'user', response: "Alice! It's been ages. I've been good, quite busy with work though. And you?" });
-conversation.canonicalDialogue.push({ speaker: 'agent', response: "Pretty much the same here. Busy, busy. I've started reading manga and watahcing anime, it's really therapeutic." });
+const getDialogue = (): string => {
+  const transcription = getTransciptionSoFar();
+  return `User: ${transcription}
+  `;
+}
+
+// const updateScreenEvents: Set<EventType> = new Set([])
+const updateScreenEvents: Set<EventType> = new Set(['responseReflex', 'cutTranscription', 'talk'])
+const updateScreen = (event: Event) => {
+  if (updateScreenEvents.has(event.eventType)) {
+    console.log(event);
+  }
+}
 
 // EVENTS
-const responseReflexEventHandler = async (conversation: Conversation): Promise<void> => {
-  const mostRecentTranscription = conversation.transcriptions[conversation.transcriptions.length - 1];
-  const userTranscriptionSinceWatermark = getTranscriptionSinceWatermark(conversation);
+const newEventHandler = (event: Event): void => {
```

Comment:
do we need some kind of random uuid kind of execution id here?

I am just thinking out loud here, but how will we propagate errors otherwise for one DAG execution?

plus what happens if multiple DAGs are triggered in parallel?

Reply:
I thought of this, but avoiding adding it for now, I'm using the timestamp as the ID, effectively